torch/csrc/export/upgrader.cpp

```diff
@@ -0,0 +1,253 @@
+  std::ostringstream error_stream;
+  error_stream << "Error in upgrader '" << upgrader_name << "' "
+               << "while upgrading from version " << from_version
+               << " to version " << from_version + 1 << ": " << error_message;
+
+  if (!problematic_object.empty()) {
+    error_stream << "\nProblematic object: " << problematic_object.dump(2);
+  }
+
+  throw std::runtime_error(error_stream.str());
+}
+
+static nlohmann::json upgradeImpl(
+    const nlohmann::json& artifact,
+    int target_version,
+    bool validate_target) {
+  auto current_artifact = artifact;
+
+  // Validate that the artifact contains required schema version information
+  if (!current_artifact.contains("schema_version")) {
+    throw std::runtime_error("Missing schema_version field in artifact");
+  }
+
+  int current_version = current_artifact["schema_version"]["major"];
+
+  // Iteratively apply upgraders until target version is reached or no more are
+  // available
+  while (current_version < target_version) {
+    // Look up upgraders for the current version
+    const auto& upgraders = getUpgrader(current_version);
+
+    if (upgraders.empty()) {
+      // No more upgraders available - stop upgrading
+      break;
+    }
+
+    // Apply all upgraders for this version in bottom-up order
+    // (deeper keypaths first to prevent parent/child conflicts)
+    for (const auto& upgrader : upgraders) {
+      // Extract the field to be upgraded using its keypath
+      auto field_to_upgrade =
+          getFieldByKeypath(current_artifact, upgrader.keypath);
+
+      // Apply the upgrade transformation
+      auto upgraded_field = upgrader.upgrade_func(field_to_upgrade);
+
+      // Update the artifact with the upgraded field
+      setFieldByKeypath(current_artifact, upgrader.keypath, upgraded_field);
+    }
+
+    // Move to the next version for potential additional upgrades
+    current_version++;
+  }
+
+  // Update schema version to reflect the final upgraded version
+  if (current_artifact["schema_version"]["major"] != current_version) {
+    current_artifact["schema_version"]["major"] = current_version;
+    // Reset minor version to 0 - the correct minor version should be set
+    // when converting the json to in memory representation of ExportedProgram
+    current_artifact["schema_version"]["minor"] = 0;
+  }
+
+  // Validate that we reached the target version if requested
+  if (validate_target && current_version != target_version) {
+    std::ostringstream error_stream;
+    error_stream
+        << "Failed to upgrade to target version " << target_version
+        << ". Final version reached: " << current_version
+        << ". This may indicate missing upgraders for intermediate versions.";
+    throw std::runtime_error(error_stream.str());
+  }
+
+  return current_artifact;
+}
+
+nlohmann::json upgrade(const nlohmann::json& artifact) {
+  return upgradeImpl(artifact, std::numeric_limits<int>::max(), false);
```

Comment:
In what case would we want to use max? I feel like we would just always want the existing schema version?

Reply:
hm but are you using this anywhere? all your test cases are specifying the version